### PR TITLE
Add Node Pilot domain

### PR DESCRIPTION
**Pull request by zaelgohary**
### Description

- Add Node Pilot domain

image

### Related Issues

- https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2909

### Checklist

- [ ] Tests included
- [x] Build pass
- [ ] Documentation
- [x] Code format and docstrings
- [x] Screenshots/Video attached (needed for UI changes)

### CODE REVIEW

File: packages/playground/src/weblets/tf_node_pilot.vue [hunk 1]
```diff
@@ -141,13 +166,34 @@ async function deploy() {
...
+
+  if (!selectionDetails.value?.domain?.enableSelectedDomain) {
+    vm[0].customDomain = selectionDetails.value?.domain?.customDomain;
+    finalize(vm);
+    return;
+  }
+
+  try {
+    layout.value.setStatus("deploy", "Preparing to deploy gateway...");
+    console.log("vm[0]", vm[0]);
```
Comment (Mahmoud-Emad) [inline packages/playground/src/weblets/tf_node_pilot.vue:181]: Remove this line if not needed

File: packages/playground/src/weblets/tf_node_pilot.vue [hunk 2]
```diff
@@ -141,13 +166,34 @@ async function deploy() {
...
+      port: 34416,
+      network: vm[0].interfaces[0].network,
+    });
+
+    finalize(vm);
+  } catch (e) {
+    layout.value.setStatus("deploy", "Rollbacking back due to fail to deploy gateway...");
+
+    await rollbackDeployment(grid!, name.value);
+    layout.value.setStatus("failed", normalizeError(e, "Failed to deploy a Casperlabs instance."));
```
Comment (Mahmoud-Emad) [inline packages/playground/src/weblets/tf_node_pilot.vue:195]:
```suggestion
    layout.value.setStatus("failed", normalizeError(e, "Failed to deploy a Node Pilot instance."));
```

File: packages/playground/src/weblets/tf_node_pilot.vue [hunk 3]
```diff
@@ -102,20 +103,41 @@ const selectionDetails = ref<SelectionDetails>();
...
+    deploymentName: name.value,
+    projectName,
+    twinId: profileManager.profile!.twinId,
+  });
+
+  const domain = selectionDetails.value?.domain?.enabledCustomDomain
+    ? selectionDetails.value.domain.customDomain
+    : subdomain + "." + selectionDetails.value?.domain?.selectedDomain?.publicConfig.domain;
+
+  let vm: any;
```
Comment (Mahmoud-Emad) [inline packages/playground/src/weblets/tf_node_pilot.vue:128]: Could we specify the type of the VM instead of using `any`?

**Review by Mahmoud-Emad — CHANGES_REQUESTED** [review]
Good job ya Zainab!

File: packages/playground/src/utils/types.ts [hunk 4]
```diff
@@ -14,3 +14,39 @@ export interface IPublicConfig {
...
+  entrypoint: string;
+  metadata?: any;
+  description?: any;
+  rootfs_size: number;
+  corex: boolean;
+  gpu: any[];
+  deploymentName: string;
+  projectName: string;
+  wireguard: string;
+}
```
Comment (Mahmoud-Emad) [inline packages/playground/src/utils/types.ts:52]: It's not good if the machine type is not defined in the grid client. If it's not defined, it should be created, and if it is defined, it needs to be exported. However, writing a `GridClient` type in the `Dashboard` is not a good approach.

**Review by Mahmoud-Emad — CHANGES_REQUESTED** [review]

File: packages/playground/src/utils/types.ts (hunk 4 above)
Comment (zaelgohary) [reply packages/playground/src/utils/types.ts]: Done.

File: packages/grid_client/src/helpers/types.ts [hunk 5]
```diff
@@ -89,3 +89,41 @@ export interface ZmachineData {
   /** The list of the GPUs */
   gpu: string[] | undefined;
 }
+
+interface VM {
+  version: number;
+  contractId: number;
```
Comment (Mahmoud-Emad) [inline packages/grid_client/src/helpers/types.ts:95]: Upon reviewing the newly created file `types.ts`, I noticed that there is an interface for the `ZOS Machine` called `ZMachineData` that supports most of the `VM` types. However, it requires some additional types.

We can add the extra types as optional fields, like this:

```ts
interface ZMachineData {
  ... // Old fields
  deploymentName: string;
  projectName: string;
  wireguard: string;
}
```

**Review by Mahmoud-Emad — CHANGES_REQUESTED** [review]

File: packages/grid_client/src/helpers/types.ts (hunk 5 above)
Comment (zaelgohary) [reply packages/grid_client/src/helpers/types.ts]: I extended VM from ZmachineData.

**Review by Mahmoud-Emad — APPROVED** [review]

**Review by 0oM4R — APPROVED** [review]
create a domain name works fine, some other flows have bugs will open issues for them

image

[Screencast from 27 يون, 2024 EEST 01:19:10 م.webm](https://github.com/threefoldtech/tfgrid-sdk-ts/assets/62248851/4e953242-b206-48f1-ad8e-d0e9ad66d624)

**Pull request merged**